File: platforms/ios/AJLibrary/AJLibrary/AJApp.swift
//
//  AJApplication.swift
//  AJ
//
//  Created by Bruno Fortunato on 08/03/16.
//  Copyright © 2016 Bruno Fortunato. All rights reserved.
//

import UIKit
import AJ
import ApplicaFramework

var __AJApp_instance: AJApp?

open class AJApp {

    open var _runtime: AJRuntime?
    open var debug = false
    open var socketUrl = "ws://localhost:3000"
    
    open class func current() -> AJApp {
        return __AJApp_instance!
    }
    
    open class func runtime() -> AJRuntime {
        return current()._runtime!
    }
    
    open class func plugin(_ name: String) -> AJPlugin {
        return current()._runtime!.get(plugin: name)
    }

    public init() {
        if (__AJApp_instance != nil) {
            fatalError("AJApp already instantiated. It's possible to have only one instance of AJApp")
        }
        
        __AJApp_instance = self
    }

    func createWebSocketRuntime() -> AJRuntime {
        return AJWebSocketRuntime(url: URL(string: socketUrl)!)
    }

    func createNativeRuntime() -> AJRuntime {
        return AJJavaScriptCoreRuntime()
    }

    open func initialize() {
        if (_runtime == nil) {
            if debug {
                _runtime = createWebSocketRuntime()
            } else {
                _runtime = createNativeRuntime()
            }
        }

    }

    open func createRoot() -> UIViewController {
        return UIViewController()
    }
    
    open func reload() {
        if let runtime = self._runtime {
            runtime.destroy()
            self._runtime = nil
        }

        initialize()
    }
    
}

File: platforms/android/App/aj/src/main/java/applica/aj/AJApp.java
package applica.aj;

import android.content.Context;

import junit.framework.Assert;

import java.net.URI;

import applica.aj.runtime.AJRuntime;
import applica.aj.runtime.javascriptcore.AJJavaScriptCoreRuntime;
import applica.aj.runtime.v8.AJV8Runtime;
import applica.aj.runtime.websocket.AJWebSocketRuntime;

/**
 * Created by bimbobruno on 21/04/16.
 */
public class AJApp {

    private static AJApp _current;

    public static AJApp current() {
        return _current;
    }

    public static AJRuntime runtime() {
        return current().getRuntime();
    }

    private Context context;
    private AJRuntime runtime;
    private boolean debug = false;
    private String socketUrl = "http://localhost:3000";

    public AJApp(Context context) {
        Assert.assertNull("AJApp already instantiated. It's possible to have only one m_instance of AJApp", _current);

        _current = this;

        this.context = context;
    }

    public void init() {
        if (debug) {
            runtime = new AJWebSocketRuntime(context, URI.create(socketUrl));
        } else {
            runtime = new AJV8Runtime(context);
        }
    }

    public void reload() {
        if (runtime != null) {
            runtime.destroy();
        }

        init();
    }

    public AJRuntime getRuntime() {
        return runtime;
    }

    //alias
    public AJRuntime rt() {
        return getRuntime();
    }

    public boolean isDebug() {
        return debug;
    }

    public void setDebug(boolean debug) {
        this.debug = debug;
    }

    public String getSocketUrl() {
        return socketUrl;
    }

    public void setSocketUrl(String socketUrl) {
        this.socketUrl = socketUrl;
    }
}
